### DISCUSSION TITLE
Is that possible persist data during local development?

### DISCUSSION THREAD

**Discussion by burmecia** [supabase/supabase#5725] [Questions] [answered]
I've followed the [local development instructions](https://supabase.com/docs/guides/local-development), but all the tables newly created in Studio are lost when supabase service stopped and started again. Can we make the data persistent between developing sessions?

**Comment by burmecia** [comment]
To be precise, 'to persist data' I mean persist data created by user in Postgres. That will be really helpful for developer as stop&start cycle will be quite often during local development.

**Reply by 0-vortex** [reply]
https://github.com/supabase/cli/tree/main/examples/tour mentions `seed.sql` - there are ways to add data through multiple files by using the `psql` from `docker exec`

**Reply by burmecia** [reply]
thanks for the info, that is an option but I think it could be better if that can be automated by a simple command line option.

**Reply by 0-vortex** [reply]
we plugged into the docker container and ran psql as one seed file was not enough: https://github.com/open-sauced/hot/blob/f91ca82a20240b255cb71869854d21985ef49427/supabase/seed.sh - do not recommend going this way at all! 

Would honestly love a better way of persisting multiple tables :<

**Comment by burmecia** [comment]
I just found a workaround to achieve this, you can use `supabase db commit` to save changes made to a migration script, and then next time restart those changes will be automatically applied.

**Reply by growthondev** [reply]
@burmecia does this process save the data as well? I thought it only saves the schema.

**Comment by jacobsmith** [comment]
I'm using `pg_dump` to dump the whole database (data included) and then put it back after restarting. It's not automated (yet - should be pretty simple), but gets around the issue of data not being present. Modeled after the documentation here: https://supabase.com/docs/guides/database#migrate-the-database

Run ALTER ROLE postgres SUPERUSER in the old project's SQL editor
Run pg_dump --clean --if-exists --quote-all-identifiers -h $OLD_DB_URL -U postgres > dump.sql from your terminal
Run ALTER ROLE postgres NOSUPERUSER in the old project's SQL editor
Run ALTER ROLE postgres SUPERUSER in the new project's SQL editor
Run psql -h $NEW_DB_URL -U postgres -f dump.sql from your terminal
Run TRUNCATE storage.objects in the new project's SQL editor
Run ALTER ROLE postgres NOSUPERUSER in the new project's SQL editor

In this instance $OLD_DB_URL and $NEW_DB_URL would be the same, just one before you shut down the supabase server and one run after.

**Comment by jacekczapiewski** [comment]
It is not a perfect solution but You can uncomment volume in the `docker-compose.yml` file (https://github.com/supabase/supabase/blob/master/docker/docker-compose.yml#L167)

**Comment by sweatybridge** [comment, marked as answer]
This is now possible in CLI with `supabase stop --backup` flag. A pg_dump will be created at `supabase/.branches/main/dump.sql` and restored on `supabase start`. We hope to make this the default behaviour for local development once most edge cases are ironed out.

Note that the `--backup` flag persists both data and _schema_ changes between development sessions. Once you are done developing a feature, you still want to commit those separately as a [new migration](https://supabase.com/docs/guides/cli/cicd-workflow#creating-a-new-migration) for schema changes and `seed.sql` for test data.

**Reply by mellowcello77** [reply]
I have been doing this for a few weeks and seem to still lose my data when switching between projects. Will look into it but my experience has been reimporting data, would love a smooth automated setup, off course we will 🥳

**Reply by saltcod** [reply]
@GaryAustin1 could you help mark this as the answer? I forget how we decided we'd do this now, apologies.

**Reply by GaryAustin1** [reply]
There is no answer for a feature request... only for questions.   Let me see if I can turn this into a question....

**Reply by Akash187** [reply]
Hi I can't find dump.sql according to you "A pg_dump will be created at supabase/.branches/main/dump.sql"

**Reply by ABSanthosh** [reply]
> > no more saved in supabase/.branches/main/dump.sql ?
> 
> Yup, your local dev data is backed up physically to a docker volume by default now.  Use `supabase stop --no-backup` if you don't want to save a backup.
> 
> You can still get a logical backup of the local database manually via pg_dump or
> ```
> supabase db dump --db-url postgres://postgres:postgres@localhost:54322/postgres
> ```
> But restoring has to be done manually via `psql -h localhost -p 54322 -U postgres -f dump.sql`.

It has been changed now. Check this [comment](https://github.com/supabase/cli/issues/150#issuecomment-1545347766) by a contributor.

**Comment by mxmzb** [comment]
can't we just plug in a localhost db (like, actual db, not from within docker) into the supabase instance?

**Reply by rienheuver** [reply]
This seems like the obvious route. I've been experimenting with local development, creating migrations with the CLI et cetera and totally expected config options for this in `config.toml`. I feel like this should be doable with the docker containers, much like it's already possible on the self-hosted supabase. What's your view on this @sweatybridge ?
